server/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -299,34 +337,164 @@ private void updateToLoadAndDrop(List<Notice> notices, Map<String, LookupExtract
+        return null;
+      } else {
+        Map<String, LookupExtractorFactoryContainer> lookupMap = jsonMapper.readValue(
+            response.getContent(),
+            LOOKUPS_ALL_REFERENCE
+        );
+        lookupMap.forEach((k, v) -> lookupBeanList.add(new LookupBean(k, null, v)));
+
+      }
+      return lookupBeanList;
+    }
+    catch (Exception e) {
+      LOG.error(e, "Error while trying to get lookup list from coordinator for tier[%s]", tier);
+      return null;
     }
   }
 
-  private void loadSnapshotAndInitStateRef()
+  /**
+   * Returns a list of lookups from the snapshot if the lookupsnapshottaker is configured. If it's not available, returns null.
+   *
+   * @return list of LookupBean objects, or null
+   */
+  @Nullable
+  private List<LookupBean> getLookupListFromSnapshot()
   {
     if (lookupSnapshotTaker != null) {
-      ImmutableMap.Builder<String, LookupExtractorFactoryContainer> builder = ImmutableMap.builder();
+      return lookupSnapshotTaker.pullExistingSnapshot();
+    }
+    return null;
+  }
 
-      final List<LookupBean> lookupBeanList = lookupSnapshotTaker.pullExistingSnapshot();
-      for (LookupBean lookupBean : lookupBeanList) {
-        LookupExtractorFactoryContainer container = lookupBean.getContainer();
+  private List<LookupBean> getLookupBeanList(Map<String, LookupExtractorFactoryContainer> lookupMap)
+  {
+    List<LookupBean> lookups = new ArrayList<>(lookupMap.size());
+    for (Map.Entry<String, LookupExtractorFactoryContainer> e : lookupMap.entrySet()) {
+      lookups.add(new LookupBean(e.getKey(), null, e.getValue()));
+    }
+    return lookups;
+  }
 
-        if (container.getLookupExtractorFactory().start()) {
-          builder.put(lookupBean.getName(), container);
-        } else {
-          throw new ISE("Failed to start lookup [%s]:[%s]", lookupBean.getName(), container);
-        }
+  private List<LookupBean> getLookupsListFromLookupConfig()
+  {
+    List<LookupBean> lookupBeanList = new ArrayList<>();
+    if (lookupConfig.getEnableLookupSyncOnStartup()) {
+      String tier = lookupListeningAnnouncerConfig.getLookupTier();
+      lookupBeanList = getLookupListFromCoordinator(tier);
+      if (lookupBeanList == null) {
+        LOG.info("Coordinator is unavailable. Loading saved snapshot instead");
+        lookupBeanList = getLookupListFromSnapshot();
       }
-
-      stateRef.set(new LookupUpdateState(builder.build(), ImmutableList.of(), ImmutableList.of()));
     } else {
-      stateRef.set(new LookupUpdateState(ImmutableMap.of(), ImmutableList.of(), ImmutableList.of()));
+      lookupBeanList = getLookupListFromSnapshot();
+    }
+    return lookupBeanList;
+  }
+
+  private void startLookups(List<LookupBean> lookupBeanList)
+  {
+    ImmutableMap.Builder<String, LookupExtractorFactoryContainer> builder = ImmutableMap.builder();
+    ListeningScheduledExecutorService executorService = MoreExecutors.listeningDecorator(
```

Comment:
Could you please rewrite it using standard Java API, `ExecutorCompletionService`?

Reply:
I see that ScheduledExecutorService is not required here. I have changed it to ListeningExecutorService. Is this sufficient or do I still need to refactor it to use ExecutorCompletionService?

Reply:
I suggest to use `ExecutorCompletionService` because it's standard Java API, which replaces Guava functionality that you used, that means that this functionality has a good chance of being deprecated and removed from Guava in the future, adding compatibility problems.

Because of this dumb Guava's compatibility policy (remove things), we better use Guava as little as possible.